sdk/containerregistry/container-registry/src/blob/containerRegistryBlobClient.ts

```diff
@@ -0,0 +1,400 @@
+      }
+
+      const bodyData = Buffer.from(body);
+      const expectedDigest = await calculateDigest(bodyData);
+
+      if (digest !== expectedDigest) {
+        throw new Error(`Docker-Content-Digest header does not match calculated digest.
+        Expected: ${expectedDigest}
+        Actual: ${digest}`);
+      }
+
+      const manifest = serializer.deserialize(Mappers.OCIManifest, JSON.parse(body), "OCIManifest");
+
+      return {
+        digest,
+        manifest,
+        manifestStream: Readable.from(bodyData),
+      };
+    } catch (e) {
+      span.setStatus({ code: SpanStatusCode.ERROR, message: e.message });
+      throw e;
+    } finally {
+      span.end();
+    }
+  }
+
+  /**
+   * Delete a manifest. Doing so effectively deletes an artifact from the registry.
+   *
+   * @param digest - the digest of the manifest to delete
+   * @param options - optional configuration used to send requests to the service
+   */
+  public async deleteManifest(digest: string, options?: DeleteManifestOptions): Promise<void> {
+    const { span, updatedOptions } = createSpan(
+      "ContainerRegistryBlobClient-deleteManifest",
+      options
+    );
+
+    try {
+      await this.client.containerRegistry.deleteManifest(
+        this.repositoryName,
+        digest,
+        updatedOptions
+      );
+    } catch (e) {
+      span.setStatus({ code: SpanStatusCode.ERROR, message: e.message });
+      throw e;
+    } finally {
+      span.end();
+    }
+  }
+
+  /**
+   * Upload an artifact blob.
+   *
+   * @param blobStreamFactory - a factory which produces a stream containing the blob data. This function may be called multiple times; each time the function is called a fresh stream should be returned.
+
+   */
+  public async uploadBlob(
+    blobStreamFactory: () => NodeJS.ReadableStream
+  ): Promise<UploadBlobResult>;
+
+  /**
+   * Upload an artifact blob.
+   *
+   * @param blobStream - the stream containing the blob data.
+   */
+  public async uploadBlob(blobStream: NodeJS.ReadableStream): Promise<UploadBlobResult>;
+
+  public async uploadBlob(
+    blobStreamOrFactory: (() => NodeJS.ReadableStream) | NodeJS.ReadableStream
+  ): Promise<UploadBlobResult> {
+    const { span } = createSpan("ContainerRegistryBlobClient-uploadBlob", undefined);
+
+    try {
+      const startUploadResult = await this.client.containerRegistryBlob.startUpload(
+        this.repositoryName
+      );
+
+      let requestBody: NodeJS.ReadableStream | Buffer;
+      let digest: string;
+
+      if (typeof blobStreamOrFactory === "function") {
+        requestBody = blobStreamOrFactory();
+        digest = await calculateDigest(blobStreamOrFactory());
+      } else {
+        requestBody = await readStreamToEnd(blobStreamOrFactory);
+        digest = await calculateDigest(requestBody);
+      }
+
+      if (!startUploadResult.location) {
+        throw new Error("startUpload did not return a location");
+      }
```

Comment:
similar thoughts for the errors in this function, these should be RestErrors, since essentially the response payload has failed deserialization.

Reply:
Just gonna assert here, there shouldn't be a scenario where the server doesn't provide the header on a successful response.

Reply:
To make the code a little cleaner, we can do assertion functions: https://github.com/Azure/azure-sdk-for-js/pull/20766#discussion_r840969905